onnxruntime/python/onnxruntime_pybind_state.cc

```diff
@@ -565,19 +596,33 @@ static AllocatorPtr GetCudaAllocator(OrtDevice::DeviceId id) {
 
   if (id_to_allocator_map.find(id) == id_to_allocator_map.end()) {
     // Use arena-based allocator
-    AllocatorCreationInfo default_memory_info(
-        [](OrtDevice::DeviceId id) {
-          return onnxruntime::make_unique<CUDAAllocator>(id, CUDA);
-        },
-        id,
-        true,
-        {cuda_mem_limit,
-         static_cast<int>(arena_extend_strategy),
-         -1, -1});
-
-    auto allocator = CreateAllocator(default_memory_info);
-
-    id_to_allocator_map.insert({id, allocator});
+    if (cuda_external_alloc != nullptr && cuda_external_free != nullptr) {
```

Comment:
Code here can also use some refactoring. 
e.g. ExternalAllocInfo can be a member in AllocatorCreationInfo, the complexity of switch between allocators should be encapsulated in the allocatormgr.cc.

Reply:
Allocatormgr is used for wrapping a cacheing layer (arena) around an allocator (device). The allocator is an input to it. I don't think it is systematically correct for allocator manager to be in the business of picking an allocator. Instead CudaExecutionProvider which exposes an allocator should be the one deciding what allocator to pick and pass it to allocator manager. Happy to talk offline if you disagree. 

---
In reply to: [568314945](https://github.com/microsoft/onnxruntime/pull/6529#discussion_r568314945) [](ancestors = 568314945)

Reply:
Discussed offline and agreed to refactor this code in CUDAEp